akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1343,7 +1343,9 @@ private[stream] object Collect {
         }
 
       private def pullIfNeeded(): Unit = {
-        if (buffer.used < parallelism && !hasBeenPulled(in)) tryPull(in)
+        if (isClosed(in) && buffer.isEmpty) completeStage()
+        else if (buffer.used < parallelism && !hasBeenPulled(in)) tryPull(in)
+        // else already pulled and waiting for next element
```

Comment:
With this change the first condition in `pushNextIfPossible()` could be simplified to:

```scala
        if (buffer.isEmpty) {
          pullIfNeeded()
```

https://github.com/johanandren/akka/blob/wip-28657-map-async-completion/akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala#L1318..L1320